### ISSUE TITLE
How is it possible for custom_vjp to break primal calculation?

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#3148]
Consider the following code:
```
    @custom_vjp
    def find_fixed_point(iterated_function, theta, x_init):
        ...  # a while_loop that does a calculation

    def _fs_fwd(iterated_function, theta, x_init):
        augmented = find_fixed_point(iterated_function, theta, x_init)
        return augmented, ()

    def _fs_bwd(memos, augmented_star_bar):
        assert False

    find_fixed_point.defvjp(_fs_fwd, _fs_bwd)
```

As can be seen, the backwards propagation of gradients is blocked by an assertion.  All that I'm doing is calculating the primals for now, but that is giving me "escaped tracers".  When I remove the `custom_vjp` decoration (and the corresponding `defvjp` call), everything works, and no escaped tracers are reported.

I thought that custom_vjp decoration wouldn't affect primal calculation when the forward pass is not doing anything except calling the base function.  How can `custom_vjp` break the primal calculation?

I'm in the process of producing a minimum working example.  I tried to guess what was going wrong, but I couldn't, so I will have to cut lines until I reach a MWE.  It's a long process, so I was hoping for some insight.

**Comment by mattjj** [comment]
> I thought that custom_vjp decoration wouldn't affect primal calculation when the forward pass is not doing anything except calling the base function.

That's my understanding too!

**Comment by mattjj** [comment]
I can't think of anything, but I'd like to help poke at a repro when you have one.

**Comment by NeilGirdhar** [comment]
@mattjj  Thanks, I really appreciate it.  I'm going to spend another day trying to make a MWE example, and if I can't, I'll just give you access to my repo and you're welcome to see how removing the decorator breaks things (mysteriously).

**Comment by NeilGirdhar** [comment]
Stayed up all night, but here we go.  If you remove the `@custom_vjp`, then the test runs fine!
```python
import dataclasses
from functools import partial

from jax import custom_vjp, jit
from jax import numpy as jnp
from jax.lax import scan, while_loop
from jax.tree_util import register_pytree_node


def dataclass(clz):
    data_clz = dataclasses.dataclass(frozen=True)(clz)
    data_fields = list(data_clz.__dataclass_fields__.keys())

    def tree_flatten(x):
        meta = ()
        data = tuple(getattr(x, name) for name in data_fields)
        return data, meta

    def tree_unflatten(meta, data):
        meta_args = ()
        data_args = tuple(zip(data_fields, data))
        kwargs = dict(meta_args + data_args)
        return data_clz(**kwargs)

    register_pytree_node(data_clz, tree_flatten, tree_unflatten)

    return data_clz


@dataclass
class EncodingConfiguration:
    code: jnp.ndarray


@dataclass
class RivalEncoding:
    weight: jnp.ndarray


@dataclass
class AugmentedState:
    current_state: jnp.ndarray
    last_state: jnp.ndarray
    iterations: int


class IteratedFunction:
    def __init__(self, *, iteration_limit, **kwargs):
        super().__init__(**kwargs)
        self.iteration_limit = iteration_limit

    @custom_vjp
    def find_fixed_point(self, theta, x_init):
        return while_loop(
            partial(self.state_needs_iteration, theta),
            partial(self.iterate_augmented, theta),
            AugmentedState(x_init, x_init, 0))

    def _fs_fwd(self, theta, x_init):
        augmented = self.find_fixed_point(theta, x_init)
        return augmented, ()

    def _fs_bwd(memos, augmented_star_bar):
        assert False

    find_fixed_point.defvjp(_fs_fwd, _fs_bwd)

    def state_needs_iteration(self, theta, augmented):
        all_close = True
        not_done = jnp.logical_not(jnp.logical_and(augmented.iterations > 0, all_close))

        if self.iteration_limit is None:
            return not_done

        return jnp.logical_and(augmented.iterations < self.iteration_limit, not_done)

    def iterate_augmented(self, theta, augmented):
        rec = augmented.current_state

        new_rec = EncodingConfiguration(rec.code + theta.weight)

        return AugmentedState(new_rec, rec, augmented.iterations + 1)

    @classmethod
    def tree_unflatten(cls, aux_data, values):
        return cls(iteration_limit=aux_data)

    def tree_flatten(self):
        return ((), self.iteration_limit)


register_pytree_node(
    IteratedFunction, IteratedFunction.tree_flatten, IteratedFunction.tree_unflatten)


def test_failure():
    element = RivalEncoding(55.0)

    @jit
    def sample_trajectory(mi):
        def ia(augmented):
            element = augmented.current_state
            eif = IteratedFunction(iteration_limit=mi)
            eif.find_fixed_point(element, EncodingConfiguration(20.0))
            return augmented

        def f(augmented, x):
            return ia(augmented), augmented.current_state

        initial_augmented = AugmentedState(element, element, 0)
        return scan(f, initial_augmented, None, 400)

    augmented, trajectory = sample_trajectory(400)
```

**Comment by NeilGirdhar** [comment]
I just figured out where the leak is: the tracer is being passed as a component of the auxiliary data.  This would have been caught automatically if JAX blocks non-hashable static arguments as I suggested in https://github.com/google/jax/issues/2813 since
- `JaxprTracer` is thankfully not hashable, and
- the auxilliary data of a PyTree-like object (one that is registered) is essentially a static argument of a jitted function.
